source/agora/script/Stack.d

```diff
@@ -0,0 +1,233 @@
+    In addition, it makes it very cheap to copy the stack as all internal
+    items are immutable anyway.
+
+*******************************************************************************/
+
+public struct Stack
+{
+    /// The actual stack
+    private SList!(const(ubyte)[]) stack;
+
+    /// The number of items on the stack
+    private ulong num_items;
+
+    /// Total used bytes for this stack
+    private size_t used_bytes;
+
+    /***************************************************************************
+
+        Checks if the provided data can be pushed to the stack.
+
+        Params:
+            data = the data to check
+
+        Returns:
+            true if data is within `MAX_STACK_ITEM_SIZE` limit and adding it
+            to the stack does not exceed the `MAX_STACK_TOTAL_SIZE` stack size.
+
+    ***************************************************************************/
+
+    public bool canPush (const(ubyte)[] data) const pure nothrow @safe @nogc
+    {
+        return data.length <= MAX_STACK_ITEM_SIZE &&
+            this.used_bytes + data.length <= MAX_STACK_TOTAL_SIZE;
+    }
+
+    /***************************************************************************
+
+        Pushes the value to the stack.
+
+        Call `canPush()` first to ensure the data can fit on the stack
+        based on the configured limits.
+
+        Params:
+            data = the data to push to the stack
+
+    ***************************************************************************/
+
+    public void push (const(ubyte)[] data) @safe nothrow
+    {
+        assert(this.canPush(data));
+        this.stack.insertFront(data);
+        this.used_bytes += data.length;
+        this.num_items++;
+    }
+
+    /***************************************************************************
+
+        Returns:
+            the top item on the stack, without popping it
+
+    ***************************************************************************/
+
+    public const(ubyte)[] peek () /*const*/ pure nothrow @safe @nogc
+    {
+        assert(!this.stack.empty());
+        return this.stack.front();
+    }
+
+    /***************************************************************************
+
+        Returns:
+            the popped value from the stack
+
+    ***************************************************************************/
+
+    public const(ubyte)[] pop () @safe nothrow
+    {
+        assert(!this.stack.empty());
+        assert(this.num_items > 0);
+        auto value = this.stack.front();
+        this.stack.removeFront();
+        this.used_bytes -= value.length;
+        this.num_items--;
+        return value;
+    }
+
+    /***************************************************************************
+
+        Returns:
+            the number of bytes used by this stack
+
+    ***************************************************************************/
+
+    public ulong usedBytes () const pure nothrow @safe @nogc
```

Comment:
This is actually not used by any other code. We have `canPush` instead. Let's remove it.